File: src/corrida.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <pthread.h>
#include <unistd.h>     // sleep
#include <time.h>       // time, srand, rand

// Cores ANSI (funciona em terminais compatíveis)
#define CYAN    "\x1b[36m"
#define BLUE    "\x1b[34m"
#define YELLOW  "\x1b[33m"
#define GREEN   "\x1b[32m"
#define RESET   "\x1b[0m"

// Estrutura de status enviada pelo carro
typedef struct {
    int volta;
    int combustivel;
    int pneus;
    int velocidade;
} status_t;

// Variáveis globais para comunicação
status_t status_global;
char comando_global[16];
int status_pronto = 0;
int comando_pronto = 0;

// Mutex e variáveis de condição
pthread_mutex_t mtx = PTHREAD_MUTEX_INITIALIZER;
pthread_cond_t cond_status = PTHREAD_COND_INITIALIZER;
pthread_cond_t cond_comando = PTHREAD_COND_INITIALIZER;

// Thread do carro
void* carro_corrida(void* arg) {
    const int voltas = 10;
    int combustivel = 100;
    int pneus = 100;
    int velocidade = 200;  // km/h média

    for (int volta = 1; volta <= voltas; volta++) {
        printf(CYAN "\n==================== VOLTA %d ====================\n" RESET, volta);

        // Simula desgaste
        combustivel -= (rand() % 6 + 5);   // 5 a 10%
        pneus      -= (rand() % 6 + 3);     // 3 a 8%

        // Prepara status
        status_t s = { volta, combustivel, pneus, velocidade };

        printf(BLUE "[Carro] 🚗 Enviando status: Combustível = %d%%, Pneus = %d%%, Velocidade = %d km/h\n" RESET,
               s.combustivel, s.pneus, s.velocidade);

        // Envia status para a equipe
        pthread_mutex_lock(&mtx);
        // Espera o slot ficar livre (status_pronto == 0)
        while (status_pronto)
            pthread_cond_wait(&cond_status, &mtx);
        status_global = s;
        status_pronto = 1;
        pthread_cond_signal(&cond_status);
        // Aguarda comando
        while (!comando_pronto)
            pthread_cond_wait(&cond_comando, &mtx);
        // Lê comando
        char cmd[16];
        strcpy(cmd, comando_global);
        comando_pronto = 0;
        pthread_cond_signal(&cond_comando);
        pthread_mutex_unlock(&mtx);

        // Executa ação conforme comando
        if (strcmp(cmd, "box") == 0) {
            printf(YELLOW "[Carro] 🏁 Entrando nos BOXES para reabastecimento e troca de pneus...\n" RESET);
            sleep(2);
            combustivel = 100;
            pneus = 100;
            printf(YELLOW "[Carro] ✅ Pronto para voltar à pista!\n" RESET);
        }
        else if (strcmp(cmd, "reduzir") == 0) {
            printf(YELLOW "[Carro] ⚠️ Reduzindo velocidade.\n" RESET);
            velocidade -= 20;
        }
        else if (strcmp(cmd, "acelerar") == 0) {
            printf(YELLOW "[Carro] 🟢 Acelerando ao máximo!\n" RESET);
            velocidade += 20;
        }
        else {
            printf(YELLOW "[Carro] ➡️ Mantendo ritmo de corrida.\n" RESET);
        }

        // Simula tempo da volta
        sleep(1);
    }

    printf(CYAN "\n🏁 [Carro] Corrida finalizada! 🏆\n" RESET);
    return NULL;
}

// Thread da equipe técnica
void* equipe_tecnica(void* arg) {
    while (1) {
        // Aguarda status do carro
        pthread_mutex_lock(&mtx);
        while (!status_pronto)
            pthread_cond_wait(&cond_status, &mtx);
        // Copia status
        status_t s = status_global;
        status_pronto = 0;
        pthread_cond_signal(&cond_status);

        // Decide comando
        char cmd[16];
        if (s.combustivel < 50 || s.pneus < 45) {
            strcpy(cmd, "box");
        }
        else if (s.velocidade > 220) {
            strcpy(cmd, "reduzir");
        }
        else if (s.velocidade < 180) {
            strcpy(cmd, "acelerar");
        }
        else {
            strcpy(cmd, "manter");
        }

        printf(GREEN "[Equipe] 📡 Recebido: Volta %d | Combustível = %d%% | Pneus = %d%% | Velocidade = %d km/h\n" RESET,
               s.volta, s.combustivel, s.pneus, s.velocidade);
        printf(GREEN "[Equipe] 🗣️ Comando: %s\n" RESET, cmd);

        // Envia comando de volta
        strcpy(comando_global, cmd);
        comando_pronto = 1;
        pthread_cond_signal(&cond_comando);
        pthread_mutex_unlock(&mtx);

        if (s.volta == 10)
            break;
    }
    return NULL;
}

int main() {
    srand(time(NULL));

    pthread_t thread_carro, thread_equipe;

    // Cria threads
    pthread_create(&thread_carro, NULL, carro_corrida, NULL);
    pthread_create(&thread_equipe, NULL, equipe_tecnica, NULL);

    // Aguarda fim
    pthread_join(thread_carro, NULL);
    pthread_join(thread_equipe, NULL);

    // Limpeza
    pthread_mutex_destroy(&mtx);
    pthread_cond_destroy(&cond_status);
    pthread_cond_destroy(&cond_comando);

    return 0;
}

File: src/corrida.py
import threading
import queue
import time
import random

# Queues de comunicação
status_queue = queue.Queue()
command_queue = queue.Queue()

# Função da thread do carro
def carro_corrida():
    voltas = 10
    combustivel = 100
    pneus = 100
    velocidade = 200  # km/h média
    for volta in range(1, voltas + 1):
        print(f"\n[Carro] Iniciando volta {volta}")
        
        # Simula desgaste
        combustivel -= random.randint(5, 10)
        pneus -= random.randint(3, 8)

        # Envia status para equipe
        status = {
            'volta': volta,
            'combustivel': combustivel,
            'pneus': pneus,
            'velocidade': velocidade
        }
        status_queue.put(status)

        # Aguarda comando da equipe
        comando = command_queue.get()
        if comando == 'box':
            print("[Carro] Entrando nos boxes para reabastecer e trocar pneus.")
            time.sleep(2)
            combustivel = 100
            pneus = 100
        elif comando == 'reduzir':
            print("[Carro] Reduzindo velocidade.")
            velocidade -= 20
        elif comando == 'acelerar':
            print("[Carro] Acelerando ao máximo!")
            velocidade += 20
        else:
            print("[Carro] Continuando normalmente.")
        
        # Simula tempo da volta
        time.sleep(1)

    print("\n[Carro] Corrida finalizada!")

# Função da thread da equipe
def equipe_tecnica():
    while True:
        status = status_queue.get()
        print(f"[Equipe] Status recebido: Volta {status['volta']} | Combustível: {status['combustivel']}% | Pneus: {status['pneus']}%")

        # Decide o comando
        if status['combustivel'] < 30 or status['pneus'] < 25:
            comando = 'box'
        elif status['velocidade'] > 220:
            comando = 'reduzir'
        elif status['velocidade'] < 180:
            comando = 'acelerar'
        else:
            comando = 'manter'
        
        # Envia comando
        print(f"[Equipe] Enviando comando: {comando.upper()}")
        command_queue.put(comando)

        if status['volta'] == 10:
            break

# Criando as threads
thread_carro = threading.Thread(target=carro_corrida)
thread_equipe = threading.Thread(target=equipe_tecnica)

# Iniciando as threads
thread_carro.start()
thread_equipe.start()

# Aguardando finalização
thread_carro.join()
thread_equipe.join()
